pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,288 @@
+			expected: []string{},
+		},
+		{
+			name:      "Both Agent Crds and external nodes present.",
+			AgentCRDs: []string{"TestAgentCRD1", "TestAgentCRD2", "TestAgentCRD3", "TestAgentCRD4"},
+			externalnodes: []*v1alpha1.ExternalNode{{ObjectMeta: metav1.ObjectMeta{Name: "TestAgentCRD3", Namespace: "ns1", Labels: map[string]string{"en": "vm2"}}},
+				{ObjectMeta: metav1.ObjectMeta{Name: "TestAgentCRD4", Namespace: "ns1", Labels: map[string]string{"en": "vm2"}}},
+			},
+			expected: []string{"TestAgentCRD3", "TestAgentCRD4"},
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			for _, crds := range tt.AgentCRDs {
+				agentCRD := new(v1beta1.AntreaAgentInfo)
+				agentCRD.Name = crds
+				_, err := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().Create(context.TODO(), agentCRD, metav1.CreateOptions{})
+				assert.NoError(t, err)
+				agentCRD.ResourceVersion = "0"
+			}
+			for _, externalnodes := range tt.externalnodes {
+
+				_, err := controller.controllerMonitor.client.CrdV1alpha1().ExternalNodes(externalnodes.Namespace).Create(context.TODO(), externalnodes, metav1.CreateOptions{})
+				if err != nil {
+					fmt.Println(err.Error())
+				}
+			}
+			time.Sleep(2 * time.Second)
+			controller.controllerMonitor.externalNodeEnabled = true
+			controller.controllerMonitor.deleteStaleAgentCRDs()
+			crds, _ := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().List(context.TODO(), metav1.ListOptions{ResourceVersion: "0"})
+			existingNames := []string{}
+			for _, crd := range crds.Items {
+				existingNames = append(existingNames, crd.Name)
+			}
+			assert.Equal(t, tt.expected, existingNames)
+		})
+
+	}
+
+}
+
+func TestSyncControllerCRD(t *testing.T) {
+	controller := newControllerMonitor([]runtime.Object{})
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	controller.informerFactory.Start(stopCh)
+	controller.crdInformerFactory.Start(stopCh)
+	controller.informerFactory.WaitForCacheSync(stopCh)
+	controller.crdInformerFactory.WaitForCacheSync(stopCh)
+	tc := []struct {
+		name          string
+		controllerCRD string
+	}{
+		{
+			name:          "ControllerCRD does not exist",
+			controllerCRD: "",
+		},
+		{
+			name:          "ControllerCRD exists and updated partially",
+			controllerCRD: "TestControllerCRD",
+		},
+	}
+	for _, tt := range tc {
+		t.Run("1", func(t *testing.T) {
+			if tt.controllerCRD == "" {
+				controller.controllerMonitor.syncControllerCRD()
+				assert.Equal(t, crdName, controller.controllerMonitor.controllerCRD.Name)
+			} else {
+				controller.controllerMonitor.controllerCRD, _ = controller.controllerMonitor.createControllerCRD(tt.controllerCRD)
+				controller.controllerMonitor.syncControllerCRD()
+				assert.NotNil(t, controller.controllerMonitor.controllerCRD)
+			}
+
+		})
+	}
+
+}
+
+func newControllerMonitor(objects []runtime.Object) *fakeController {
```

Comment:
Is there any case you passed objects to `newControllerMonitor`? if not, then there is unnecessary to add the parameter `objects []runtime.Object`.